src/main/java/com/epam/aidial/core/token/TokenStatsTracker.java

```diff
@@ -28,45 +140,79 @@ public Future<TokenUsage> getTokenStats(ProxyContext context) {
+        return registeredTrackers;
+    }
+
+    private class TraceContext {
         Map<String, TokenStats> spans = new HashMap<>();
 
-        synchronized void addSpan(ProxyContext context) {
+        synchronized Future<Void> addSpan(ProxyContext context) {
             String spanId = context.getSpanId();
             String parentSpanId = context.getParentSpanId();
-            TokenStats tokenStats = new TokenStats(new TokenUsage(), Promise.promise(), new ArrayList<>());
+            TokenStats tokenStats = new TokenStats(new TokenUsage(), new HashMap<>());
             spans.put(spanId, tokenStats);
             if (parentSpanId != null) {
-                TokenStats parent = spans.get(parentSpanId);
-                if (parent != null) {
-                    parent.children.add(tokenStats.promise.future());
-                }
+                StartSpanEvent event = new StartSpanEvent(trackerId, context.getTraceId(), parentSpanId, context.getSpanId());
+                return publish(Event.START_SPAN, event);
             }
+            return Future.succeededFuture();
+        }
+
+        synchronized Promise<TokenUsage> updateSpan(StartSpanEvent event) {
+            TokenStats tokenStats = spans.get(event.parentSpanId);
+            if (tokenStats == null) {
+                return null;
+            }
+            Promise<TokenUsage> promise = Promise.promise();
+            tokenStats.children.put(event.spanId, promise);
+            return promise;
+        }
+
+        synchronized Promise<TokenUsage> updateSpan(EndSpanEvent event) {
+            TokenStats tokenStats = spans.get(event.parentSpanId);
+            if (tokenStats == null) {
+                return null;
+            }
+            Promise<TokenUsage> promise = tokenStats.children.get(event.spanId);
+            if (promise == null) {
+                return null;
+            }
+            promise.complete(event.tokenUsage);
+            return promise;
         }
 
-        synchronized void endSpan(ProxyContext context) {
+        synchronized Future<Void> endSpan(ProxyContext context) {
             String spanId = context.getSpanId();
-            TokenStats tokenStats = spans.get(spanId);
-            if (tokenStats != null) {
-                tokenStats.promise.complete(context.getTokenUsage());
+            spans.remove(spanId);
+            if (spans.isEmpty()) {
+                TokenStatsTracker.this.traceIdToContext.remove(context.getTraceId());
             }
+            String parentSpanId = context.getParentSpanId();
+            if (parentSpanId != null) {
+                EndSpanEvent event = new EndSpanEvent(trackerId, context.getTraceId(), parentSpanId, context.getSpanId(), context.getTokenUsage());
+                return publish(Event.END_SPAN, event);
+            }
+            return Future.succeededFuture();
         }
 
         synchronized Future<TokenUsage> getStats(ProxyContext context) {
```

Comment:
please add javadoc/comment how and when this method should be called

Reply:
fixed